File: test_factory.py
"""Tests for factory.py design_factory function"""

import re

from factory import design_factory
from recipes import Purity, get_recipes_for


def test_get_recipes_for():
    """Test recipe lookup"""
    recipes = get_recipes_for("Iron Plate")
    assert recipes, "Should find recipes for Iron Plate"
    assert 20.0 in recipes, "Should include 20/min recipe"
    print("PASS: get_recipes_for")


def test_design_factory_with_input():
    """Test factory design with raw material input"""
    factory = design_factory(
        outputs={"Iron Plate": 100}, inputs=[("Iron Ore", 500)], mines=[]
    )

    assert factory.network is not None
    assert len(factory.network.body) > 0

    # Check that machines were created
    machines = [n for n in factory.network.body if "Machine_" in str(n)]
    assert len(machines) > 0, "Should create machine nodes"

    print(f"PASS: design_factory with input ({len(machines)} machines)")


def test_design_factory_with_mine():
    """Test factory design with mining node"""
    factory = design_factory(
        outputs={"Iron Plate": 100}, inputs={}, mines=[("Iron Ore", Purity.PURE)]
    )

    assert factory.network is not None
    assert len(factory.network.body) > 0

    # Check that mine was created (in source string since body is implementation detail)
    source = factory.network.source
    has_mine = "Mine_" in source or "Miner" in source
    assert has_mine, "Should create mine node"

    print("PASS: design_factory with mine")


def test_balancers_included():
    """Test that splitters and mergers are included in the graph"""
    # Create scenario that needs balancing
    factory = design_factory(
        outputs={"Iron Plate": 100}, inputs=[("Iron Ore", 500)], mines=[]
    )

    source = factory.network.source

    # Check for splitters or mergers (may not always be needed depending on recipe)
    has_splitters = bool(re.search(r"S\d+", source))
    has_mergers = bool(re.search(r"M\d+", source))

    # At least should have proper routing
    has_edges = "->" in source
    assert has_edges, "Should have edges connecting nodes"

    print(f"PASS: balancers check (splitters={has_splitters}, mergers={has_mergers})")


def test_material_flow_labels():
    """Test that edges have material and flow labels"""
    factory = design_factory(
        outputs={"Iron Plate": 100}, inputs=[("Iron Ore", 500)], mines=[]
    )

    source = factory.network.source

    # Check for edges in the graph - labels may be in various formats
    # Look for label attribute in graphviz format
    has_edges = "->" in source
    assert has_edges, "Should have edges connecting nodes"
    
    # Check for label attributes (graphviz format uses label=)
    has_labels = "label=" in source
    assert has_labels, "Should have label attributes on nodes or edges"

    print(f"PASS: material flow labels")


def test_no_recipe_error():
    """Test that missing raw materials are auto-generated instead of erroring"""
    # Old behavior: would raise an error
    # New behavior: auto-generates required raw materials
    factory = design_factory(
        outputs={"Iron Plate": 100},
        inputs=[],
        mines=[],  # No iron ore provided  # No mines
    )
    # Should succeed and auto-generate Iron Ore input
    assert factory.network is not None
    source = factory.network.source
    assert "Iron Ore" in source
    print("PASS: auto-generates missing raw materials")


def test_factory_dataclass():
    """Test that Factory dataclass is properly populated"""
    outputs = {"Iron Plate": 100}
    inputs = [("Iron Ore", 500)]
    mines = []

    factory = design_factory(outputs, inputs, mines)

    # Check that all requested outputs are present (may include byproducts)
    for material, amount in outputs.items():
        assert material in factory.outputs, f"Missing requested output: {material}"
        assert (
            factory.outputs[material] >= amount
        ), f"Insufficient output for {material}: {factory.outputs[material]} < {amount}"

    assert factory.inputs == inputs
    assert factory.mines == mines
    assert factory.network is not None

    print("PASS: Factory dataclass")


def test_auto_raw_materials():
    """Test automatic raw material detection"""
    # Request concrete without specifying limestone input
    factory = design_factory(
        outputs={"Concrete": 480},
        inputs=[],  # No inputs - should auto-detect limestone need
        mines=[],
    )

    assert factory.network is not None
    source = factory.network.source

    # Should have auto-generated limestone input
    assert "Limestone" in source
    assert "auto" in source.lower()

    # Should have constructor machines
    assert "Constructor" in source

    print("PASS: Auto raw materials detection")


def test_complex_production_chain():
    """Test complex production with multiple steps"""
    # Create a more complex factory with multiple production steps
    factory = design_factory(
        outputs={"Iron Plate": 200, "Copper Ingot": 100},
        inputs=[],
        mines=[("Iron Ore", Purity.NORMAL), ("Copper Ore", Purity.NORMAL)],
    )

    assert factory.network is not None
    source = factory.network.source

    # Should have mine nodes
    assert "Mine" in source or "Miner" in source

    # Should have multiple machines
    assert "Smelter" in source or "Machine" in source

    print("PASS: Complex production chain")


def test_simple_single_machine():
    """Test very simple case with single machine"""
    # Simple case that should create direct connections
    factory = design_factory(
        outputs={"Iron Plate": 30},
        inputs=[("Iron Ore", 30)],
        mines=[],
    )

    assert factory.network is not None
    assert len(factory.network.source) > 0

    print("PASS: Simple single machine case")


def test_byproduct_handling():
    """Test factory that produces byproducts"""
    # Some recipes produce multiple outputs (byproducts)
    # For example, fuel production can produce polymer resin as byproduct
    factory = design_factory(
        outputs={"Fuel": 100},
        inputs=[],
        mines=[],
    )

    assert factory.network is not None
    source = factory.network.source
    
    # Should create a valid graph with the primary output
    assert "Fuel" in source

    print("PASS: Byproduct handling")


def test_intermediate_and_final_output():
    """Test when an intermediate product is also a final output"""
    # Request both Iron Ingot and Iron Plate
    # Iron Ingot will be both produced for Iron Plate and output separately
    factory = design_factory(
        outputs={"Iron Ingot": 50, "Iron Plate": 30},
        inputs=[],
        mines=[("Iron Ore", Purity.NORMAL)],
    )

    assert factory.network is not None
    source = factory.network.source
    
    # Both should be in the output
    assert "Iron Ingot" in source
    assert "Iron Plate" in source
    
    print("PASS: Intermediate and final output")

File: test-factory.js
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { Factory, design_factory } from './factory.js';
import { Purity, get_recipes_for } from './recipes.js';

describe('Factory', () => {
    it('Factory class can be instantiated', () => {
        const factory = new Factory(null, [], {}, []);
        assert.ok(factory != null);
    });

    it('Factory class has correct properties', () => {
        const network = { test: 'network' };
        const inputs = [['Iron Ore', 100]];
        const outputs = { 'Iron Plate': 50 };
        const mines = [['Iron Ore', Purity.NORMAL]];
        
        const factory = new Factory(network, inputs, outputs, mines);
        
        assert.strictEqual(factory.network, network);
        assert.strictEqual(factory.inputs, inputs);
        assert.strictEqual(factory.outputs, outputs);
        assert.strictEqual(factory.mines, mines);
    });

    it('get_recipes_for works for Iron Plate', () => {
        const recipes = get_recipes_for("Iron Plate");
        assert.ok(recipes != null);
        assert.ok(20.0 in recipes);
    });

    it('design_factory with raw material input', async () => {
        const factory = await design_factory(
            { 'Iron Plate': 100 },
            [['Iron Ore', 500]],
            []
        );

        assert.ok(factory.network != null);
        assert.ok(factory.network.source.length > 0);
        
        // Check that machines were created
        const source = factory.network.source;
        assert.ok(source.includes('Machine_') || source.includes('Smelter'));
    });

    it('design_factory with mining node', async () => {
        const factory = await design_factory(
            { 'Iron Plate': 100 },
            [],
            [['Iron Ore', Purity.PURE]]
        );

        assert.ok(factory.network != null);
        assert.ok(factory.network.source.length > 0);

        // Check that mine was created
        const source = factory.network.source;
        const hasMine = source.includes('Mine_') || source.includes('Miner');
        assert.ok(hasMine);
    });

    it('auto-generates missing raw materials', async () => {
        const factory = await design_factory(
            { 'Iron Plate': 100 },
            [],
            []  // No inputs or mines
        );

        assert.ok(factory.network != null);
        const source = factory.network.source;
        assert.ok(source.includes('Iron Ore'));
        assert.ok(source.toLowerCase().includes('auto'));
    });

    it('Factory dataclass properly populated', async () => {
        const outputs = { 'Iron Plate': 100 };
        const inputs = [['Iron Ore', 500]];
        const mines = [];

        const factory = await design_factory(outputs, inputs, mines);

        // Check that all requested outputs are present
        for (const [material, amount] of Object.entries(outputs)) {
            assert.ok(material in factory.outputs, `Missing requested output: ${material}`);
            assert.ok(
                factory.outputs[material] >= amount,
                `Insufficient output for ${material}: ${factory.outputs[material]} < ${amount}`
            );
        }

        assert.deepStrictEqual(factory.inputs, inputs);
        assert.deepStrictEqual(factory.mines, mines);
        assert.ok(factory.network != null);
    });

    it('automatic raw material detection', async () => {
        const factory = await design_factory(
            { 'Concrete': 480 },
            [],  // No inputs - should auto-detect limestone need
            []
        );

        assert.ok(factory.network != null);
        const source = factory.network.source;

        // Should have auto-generated limestone input
        assert.ok(source.includes('Limestone'));
        assert.ok(source.toLowerCase().includes('auto'));

        // Should have constructor machines
        assert.ok(source.includes('Constructor'));
    });

    it('lowercase concrete output generation', async () => {
        const factory = await design_factory(
            { 'concrete': 480 },
            [],  // No inputs - should auto-detect limestone need
            []
        );

        assert.ok(factory.network != null);
        const source = factory.network.source;

        // Should have auto-generated limestone input
        assert.ok(source.includes('Limestone'));
        assert.ok(source.toLowerCase().includes('auto'));

        // Should have constructor machines
        assert.ok(source.includes('Constructor'));
        
        // Should successfully generate factory despite lowercase input
        assert.ok(source.length > 0);
    });

    it('complex production chain', async () => {
        const factory = await design_factory(
            { 'Iron Plate': 200, 'Copper Ingot': 100 },
            [],
            [['Iron Ore', Purity.NORMAL], ['Copper Ore', Purity.NORMAL]]
        );

        assert.ok(factory.network != null);
        const source = factory.network.source;

        // Should have mine nodes
        assert.ok(source.includes('Mine') || source.includes('Miner'));

        // Should have multiple machines
        assert.ok(source.includes('Smelter') || source.includes('Machine'));
    });

    it('simple single machine case', async () => {
        const factory = await design_factory(
            { 'Iron Plate': 30 },
            [['Iron Ore', 30]],
            []
        );

        assert.ok(factory.network != null);
        assert.ok(factory.network.source.length > 0);
    });

    it('byproduct handling', async () => {
        const factory = await design_factory(
            { 'Fuel': 100 },
            [],
            []
        );

        assert.ok(factory.network != null);
        const source = factory.network.source;
        
        // Should create a valid graph with the primary output
        assert.ok(source.includes('Fuel'));
    });

    it('intermediate and final output', async () => {
        const factory = await design_factory(
            { 'Iron Ingot': 50, 'Iron Plate': 30 },
            [],
            [['Iron Ore', Purity.NORMAL]]
        );

        assert.ok(factory.network != null);
        const source = factory.network.source;
        
        // Both should be in the output
        assert.ok(source.includes('Iron Ingot'));
        assert.ok(source.includes('Iron Plate'));
    });

    it('Computer factory graph properties', async () => {
        const factory = await design_factory(
            { 'Computer': 1 },
            [],
            []
        );

        assert.ok(factory.network != null);
        assert.ok(factory.network.nodes != null);
        
        const nodes = factory.network.nodes;
        const circularShapes = ['circle', 'oval', 'ellipse', 'doublecircle', 'point'];
        
        // Check all nodes
        for (const node of nodes) {
            // Check node is not circular (shape must be set and not circular, since default is oval)
            assert.ok(node.shape, `Node ${node.id} must have shape set (default is oval which is circular)`);
            const isCircular = circularShapes.includes(node.shape.toLowerCase());
            assert.ok(!isCircular, `Node ${node.id} should not have circular shape, got ${node.shape}`)
            
            // Check node has a border (peripheries != 0 and style != invisible)
            if (node.peripheries !== undefined) {
                assert.ok(node.peripheries !== 0 && node.peripheries !== '0', 
                    `Node ${node.id} should have a border (peripheries=${node.peripheries})`);
            }
            if (node.style) {
                const hasNoBorder = node.style === 'invis' || node.style === 'invisible' || 
                                   node.style.includes('invis');
                assert.ok(!hasNoBorder, `Node ${node.id} should have a border (style=${node.style})`);
            }
            
            // Check diamond nodes have empty label
            if (node.shape && node.shape.toLowerCase() === 'diamond') {
                const hasEmptyLabel = !node.label || node.label === '' || node.label === '""';
                assert.ok(hasEmptyLabel, 
                    `Node ${node.id} with diamond shape should have empty label, got "${node.label}"`);
            }
        }
    });

    it('disableBalancers parameter defaults to false (uses balancers)', async () => {
        const factory = await design_factory(
            { 'Iron Plate': 100 },
            [['Iron Ore', 200], ['Iron Ore', 200]],
            []
        );

        assert.ok(factory.network != null);
        const source = factory.network.source;
        
        // With balancers enabled (default), should have splitter/merger nodes (diamond shapes)
        assert.ok(source.includes('shape=diamond') || source.includes('shape="diamond"'), 
            'Should have diamond-shaped balancer nodes when disableBalancers=false');
    });

    it('disableBalancers=true creates hub nodes instead of balancers', async () => {
        const factory = await design_factory(
            { 'Iron Plate': 100 },
            [['Iron Ore', 200], ['Iron Ore', 200]],
            [],
            null,  // enablementSet
            null,  // economy
            1.0,   // inputCostsWeight
            0.0,   // machineCountsWeight
            1.0,   // powerConsumptionWeight
            false, // designPower
            true   // disableBalancers
        );

        assert.ok(factory.network != null);
        const source = factory.network.source;
        
        // With balancers disabled, should have circular hub nodes
        assert.ok(source.includes('shape=circle') || source.includes('shape="circle"'),
            'Should have circular hub nodes when disableBalancers=true');
        
        // Should have a hub for Iron Ore material
        assert.ok(source.includes('Iron_Ore_Hub'),
            'Should have Iron_Ore_Hub node');
    });

    it('disableBalancers=true skips electricity routing', async () => {
        const factory = await design_factory(
            { 'Iron Plate': 30 },
            [['Iron Ore', 30]],
            [],
            null,
            null,
            1.0,
            0.0,
            1.0,
            false,
            true  // disableBalancers
        );

        assert.ok(factory.network != null);
        const source = factory.network.source;
        
        // Should successfully generate a factory with disableBalancers
        assert.ok(source.length > 0);
        
        // Should have the expected machines
        assert.ok(source.includes('Machine') || source.includes('Smelter'));
        
        // Electricity (MWm) should not have hub nodes
        assert.ok(!source.includes('MWm_Hub'),
            'Should not create hub for electricity');
    });

    it('disableBalancers=true with multiple materials creates separate hubs', async () => {
        const factory = await design_factory(
            { 'Iron Plate': 50, 'Copper Ingot': 50 },
            [['Iron Ore', 100], ['Iron Ore', 100], ['Copper Ore', 100], ['Copper Ore', 100]],
            [],
            null,
            null,
            1.0,
            0.0,
            1.0,
            false,
            true  // disableBalancers
        );

        assert.ok(factory.network != null);
        const source = factory.network.source;
        
        // Should have hub nodes for both materials
        assert.ok(source.includes('Iron_Ore_Hub'),
            'Should have Iron_Ore_Hub node');
        assert.ok(source.includes('Copper_Ore_Hub'),
            'Should have Copper_Ore_Hub node');
        
        // Both hubs should be circular
        const ironHubMatch = source.match(/Iron_Ore_Hub[^\n]*\n[^\[]*\[[^\]]*shape=(?:")?circle(?:")?\b/);
        const copperHubMatch = source.match(/Copper_Ore_Hub[^\n]*\n[^\[]*\[[^\]]*shape=(?:")?circle(?:")?\b/);
        
        assert.ok(ironHubMatch || source.includes('Iron_Ore_Hub') && source.includes('shape=circle'),
            'Iron hub should be circular');
        assert.ok(copperHubMatch || source.includes('Copper_Ore_Hub') && source.includes('shape=circle'),
            'Copper hub should be circular');
    });

    it('disableBalancers=false uses balancer network for multiple sources', async () => {
        const factory = await design_factory(
            { 'Iron Plate': 100 },
            [['Iron Ore', 200], ['Iron Ore', 200]],
            [],
            null,
            null,
            1.0,
            0.0,
            1.0,
            false,
            false  // disableBalancers (explicit false)
        );

        assert.ok(factory.network != null);
        const source = factory.network.source;
        
        // Should have diamond-shaped splitter/merger nodes from balancer.js
        assert.ok(source.includes('shape=diamond') || source.includes('shape="diamond"'),
            'Should have diamond-shaped balancer nodes');
        
        // Should NOT have hub nodes
        assert.ok(!source.includes('_Hub'),
            'Should not have hub nodes when balancers are enabled');
    });
});
